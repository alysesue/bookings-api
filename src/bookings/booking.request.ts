export class BookingRequest {
	public startDateTime: Date;
}
