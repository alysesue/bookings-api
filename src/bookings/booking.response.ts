import {BookingStatus} from "../models";

export class BookingResponse {

	public id: number;
	public status: BookingStatus;

	public startDateTime: Date;
	public sessionDurationInMinutes: number;
}
