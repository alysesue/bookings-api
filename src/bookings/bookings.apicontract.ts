export class BookingAcceptRequest {
	public serviceProviderId: number;
}

export class BookingRequest {
	public startDateTime: Date;
	public endDateTime: Date;
	public serviceProviderId?: number;
	public outOfSlotBooking?: boolean;
}

export class BookingResponse {
	public id: number;
	public status: number;
	public startDateTime: Date;
	public endDateTime: Date;
	public sessionDurationInMinutes: number;
	public serviceId: number;
	public serviceName: string;
	public serviceProviderId?: number;
	public serviceProviderName?: string;
}

export class BookingSearchRequest {
	public serviceId?: number;
	public serviceProviderId?: number;
	public status?: number;
	public from: Date;
	public to: Date;

	constructor(from: Date, to: Date, status?: number, serviceId?: number, serviceProviderId?: number) {
		this.serviceId = serviceId;
		this.serviceProviderId = serviceProviderId;
		this.status = status;
		this.from = from;
		this.to = to;
	}
}

export class BookingProviderResponse {
	public id: number;
	public name: string;
}
