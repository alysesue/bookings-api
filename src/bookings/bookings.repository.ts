import { InRequestScope } from "typescript-ioc";
import { Between, FindConditions, FindManyOptions, InsertResult } from "typeorm";
import { Booking } from "../models";
import { RepositoryBase } from "../core/repository";

@InRequestScope
export class BookingsRepository extends RepositoryBase<Booking> {
	constructor() {
		super(Booking);
	}

	public async getBooking(id: number): Promise<Booking> {
		const repository = await this.getRepository();
		return repository.findOne(id, { relations: ['_service', '_serviceProvider'] });
	}

	public async save(booking: Booking): Promise<InsertResult> {
		const repository = await this.getRepository();
		return repository.insert(booking);
	}

	public async update(booking: Booking): Promise<Booking> {
		const repository = await this.getRepository();
		return repository.save(booking);
	}

	public async search(searchRequest: {
		serviceId?: number,
		serviceProviderId?: number,
		status?: number,
		from: Date,
		to: Date
	}): Promise<Booking[]> {
		const repository = await this.getRepository();

		const findConditions: FindConditions<Booking> = {};
		findConditions['_startDateTime'] = Between(searchRequest.from, searchRequest.to);
		if (searchRequest.status) {
			findConditions['_status'] = searchRequest.status;
		}
		if (searchRequest.serviceId) {
			findConditions['_serviceId'] = searchRequest.serviceId;
		}
		if (searchRequest.serviceProviderId) {
			findConditions['_serviceProviderId'] = searchRequest.serviceProviderId;
		}

		const findManyOptions: FindManyOptions<Booking> = { where: [findConditions], relations: ['_service', '_serviceProvider'] };
		findManyOptions['order'] = {};
		findManyOptions['order']['_id'] = 'DESC';

		return repository.find(findManyOptions);
	}
}
