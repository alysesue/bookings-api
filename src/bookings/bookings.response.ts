import { Booking } from "../models";

export class BookingsResponse {
	protected bookings: Booking[];

	constructor(users: Booking[]) {
		this.bookings = users;
	}
}
