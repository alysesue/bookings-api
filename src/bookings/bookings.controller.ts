import { Inject } from "typescript-ioc";
import { Body, Controller, Get, Header, Path, Post, Query, Route, Security, SuccessResponse, Tags } from "tsoa";
import { Booking, ServiceProvider } from "../models";
import {
	BookingAcceptRequest,
	BookingProviderResponse,
	BookingRequest,
	BookingResponse,
	BookingSearchRequest
} from "./bookings.apicontract";
import { BookingsService } from "./bookings.service";
import { TimeslotsService } from "../timeslots/timeslots.service";

@Route("v1/bookings")
@Tags('Bookings')
export class BookingsController extends Controller {
	@Inject
	private bookingsService: BookingsService;

	@Inject
	private timeslotService: TimeslotsService;

	private static mapDataModels(bookings: Booking[]): BookingResponse[] {
		return bookings?.map(BookingsController.mapDataModel);
	}

	private static mapDataModel(booking: Booking): BookingResponse {
		return {
			id: booking.id,
			status: booking.status,
			sessionDurationInMinutes: booking.sessionDurationInMinutes,
			startDateTime: booking.startDateTime,
			endDateTime: booking.getSessionEndTime(),
			serviceId: booking.serviceId,
			serviceName: booking.service?.name,
			serviceProviderId: booking.serviceProviderId,
			serviceProviderName: booking.serviceProvider?.name,
			requestedAt: booking.createdAt,
		} as BookingResponse;
	}

	private static mapProvider(provider: ServiceProvider): BookingProviderResponse {
		return {
			id: provider.id,
			name: provider.name
		} as BookingProviderResponse;
	}

	@Post()
	@SuccessResponse(201, 'Created')
	@Security("service")
	public async postBooking(@Body() bookingRequest: BookingRequest, @Header("x-api-service") serviceId: number): Promise<any> {
		bookingRequest.outOfSlotBooking = false;
		const booking = await this.bookingsService.save(bookingRequest, serviceId);
		return BookingsController.mapDataModel(booking);
	}

	@Post('admin')
	@SuccessResponse(201, 'Created')
	@Security("service")
	public async postBookingOutOfSlot(@Body() bookingRequest: BookingRequest, @Header("x-api-service") serviceId: number): Promise<any> {
		bookingRequest.outOfSlotBooking = true;
		const booking = await this.bookingsService.save(bookingRequest, serviceId);
		return BookingsController.mapDataModel(booking);
	}

	@Post('{bookingId}/accept')
	@SuccessResponse(204, 'Accepted')
	public async acceptBooking(@Path() bookingId: number, @Body() acceptRequest: BookingAcceptRequest): Promise<any> {
		await this.bookingsService.acceptBooking(bookingId, acceptRequest);
	}

	@Post('{bookingId}/cancel')
	@SuccessResponse(204, 'Cancelled')
	public async cancelBooking(@Path() bookingId: number): Promise<any> {
		await this.bookingsService.cancelBooking(bookingId);
	}

	@Get('')
	@SuccessResponse(200, "Ok")
	@Security("optional-service")
	public async getBookings(
		@Query() from: Date,
		@Query() to: Date,
		@Query() status?: number,
		@Header("x-api-service") serviceId?: number): Promise<BookingResponse[]> {

		const searchQuery = new BookingSearchRequest(from, to, status, serviceId);
		const bookings = await this.bookingsService.searchBookings(searchQuery);
		return BookingsController.mapDataModels(bookings);
	}

	@Get('{bookingId}')
	@SuccessResponse(200, 'Ok')
	public async getBooking(@Path() bookingId: number): Promise<any> {
		const booking = await this.bookingsService.getBooking(bookingId);
		return BookingsController.mapDataModel(booking);
	}

	@Get('{bookingId}/providers')
	@SuccessResponse(200, 'Ok')
	public async getBookingProviders(@Path() bookingId: number): Promise<any> {
		const booking = await this.bookingsService.getBooking(bookingId);

		const timeslotEntry = await this.timeslotService.getAvailableProvidersForTimeslot(booking.startDateTime, booking.getSessionEndTime(), booking.serviceId);
		return timeslotEntry.availableServiceProviders.map(BookingsController.mapProvider) || [];
	}
}
