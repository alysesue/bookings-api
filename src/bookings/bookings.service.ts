import { ErrorCodeV2, MOLErrorV2 } from "mol-lib-api-contract";
import { Inject, InRequestScope } from "typescript-ioc";
import { Booking, BookingStatus } from "../models";
import { BookingsRepository } from "./bookings.repository";
import { BookingAcceptRequest, BookingRequest, BookingSearchRequest } from "./bookings.apicontract";
import { TimeslotsService } from '../timeslots/timeslots.service';
import { CalendarsService } from '../calendars/calendars.service';
import { DateHelper } from "../infrastructure/dateHelper";
import { ServiceProvidersRepository } from "../serviceProviders/serviceProviders.repository";
import { intersectsDateTimeSpan } from "../tools/timeSpan";

@InRequestScope
export class BookingsService {
	@Inject
	private bookingsRepository: BookingsRepository;
	@Inject
	private calendarsService: CalendarsService;
	@Inject
	private timeslotsService: TimeslotsService;
	@Inject
	private serviceProviderRepo: ServiceProvidersRepository;

	public formatEventId(event: string): string {
		return event.split("@")[0];
	}

	private async createBooking(bookingRequest: BookingRequest, serviceId: number): Promise<Booking> {
		const duration = Math.floor(DateHelper.DiffInMinutes(bookingRequest.endDateTime, bookingRequest.startDateTime));
		if (duration <= 0) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage('End time for booking must be greater than start time');
		}

		if (bookingRequest.serviceProviderId) {
			const provider = await this.serviceProviderRepo.getServiceProvider({ id: bookingRequest.serviceProviderId });
			if (!provider) {
				throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Service provider '${bookingRequest.serviceProviderId}' not found`);
			}
		}

		return Booking.create(
			serviceId,
			bookingRequest.startDateTime,
			duration,
			bookingRequest.serviceProviderId,
			bookingRequest.refId);
	}

	public async getBooking(bookingId: number): Promise<Booking> {
		const booking = await this.bookingsRepository.getBooking(bookingId);
		if (!booking) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_NOT_FOUND).setMessage(`Booking ${bookingId} not found`);
		}
		return booking;
	}

	public async save(bookingRequest: BookingRequest, serviceId: number): Promise<Booking> {
		// Potential improvement: each [serviceId, bookingRequest.startDateTime, bookingRequest.endDateTime] save method call should be executed serially.
		// Method calls with different services, or timeslots should still run in parallel.
		const booking = await this.createBooking(bookingRequest, serviceId);
		const {startDateTime, endDateTime, serviceProviderId} = bookingRequest;

		if(!bookingRequest.outOfSlotBooking) {
			await this.validateTimeSlot(booking);
		} else {
			const acceptedBookings = await this.timeslotsService.getAcceptedBookings(startDateTime, endDateTime, serviceId, serviceProviderId);

			for (const item of acceptedBookings) {
				const intersects = intersectsDateTimeSpan({start: booking.startDateTime, end: booking.getSessionEndTime()}, item.startDateTime, item.getSessionEndTime());
				if (intersects) {
					throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Booking request not valid as it overlaps another accepted booking`);
				}
			}
		}

		await this.bookingsRepository.save(booking);
		return this.getBooking(booking.id);
	}

	public async cancelBooking(bookingId: number): Promise<Booking> {
		const booking = await this.getBookingForCancelling(bookingId);
		const eventCalId = booking.eventICalId;
		if (booking.status === BookingStatus.Accepted) {
			const provider = await this.serviceProviderRepo.getServiceProvider({ id: booking.serviceProviderId });
			if (!provider) {
				throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Service provider '${booking.serviceProviderId}' not found`);

			}
			await this.calendarsService.deleteCalendarEvent(provider.calendar, this.formatEventId(eventCalId));
		}
		booking.status = BookingStatus.Cancelled;
		await this.bookingsRepository.update(booking);

		return booking;
	}

	public async acceptBooking(bookingId: number, acceptRequest: BookingAcceptRequest): Promise<Booking> {
		const booking = await this.getBookingForAccepting(bookingId);

		const provider = await this.serviceProviderRepo.getServiceProvider({ id: acceptRequest.serviceProviderId });
		if (!provider) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Service provider '${acceptRequest.serviceProviderId}' not found`);
		}
		const timeslotEntry = await this.timeslotsService.getAvailableProvidersForTimeslot(booking.startDateTime, booking.getSessionEndTime(), booking.serviceId);
		const isProviderAvailable = timeslotEntry.availableServiceProviders.filter(e => e.id === acceptRequest.serviceProviderId).length > 0;
		if (!isProviderAvailable) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Service provider '${acceptRequest.serviceProviderId}' is not available for this booking.`);
		}

		const eventICalId = await this.calendarsService.createCalendarEvent(booking, provider.calendar);

		booking.status = BookingStatus.Accepted;
		booking.serviceProvider = provider;
		booking.eventICalId = eventICalId;
		booking.acceptedAt = new Date();

		await this.bookingsRepository.update(booking);

		return booking;
	}

	private async validateTimeSlot(booking: Booking) {
		const timeslotEntry = await this.timeslotsService.getAvailableProvidersForTimeslot(booking.startDateTime, booking.getSessionEndTime(), booking.serviceId, booking.serviceProviderId);

		if (timeslotEntry.availabilityCount < 1) {
			const errorMessage = booking.serviceProviderId ? "The service provider is not available for this timeslot"
				: "No available service providers for this timeslot";
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(errorMessage);
		}
	}

	private async getBookingForAccepting(bookingId: number): Promise<Booking> {
		const booking = await this.getBooking(bookingId);
		if (booking.status !== BookingStatus.PendingApproval) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Booking ${bookingId} is in invalid state for accepting`);
		}
		return booking;
	}

	private async getBookingForCancelling(bookingId: number): Promise<Booking> {
		const booking = await this.getBooking(bookingId);
		if (booking.status === BookingStatus.Cancelled || booking.startDateTime < new Date()) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setMessage(`Booking ${bookingId} is in invalid state for cancelling`);

		}
		return booking;
	}
	public async searchBookings(searchRequest: BookingSearchRequest): Promise<Booking[]> {
		return await this.bookingsRepository.search(searchRequest);
	}
}
