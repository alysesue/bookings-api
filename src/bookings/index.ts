import {BookingsController} from "./bookings.controller";
import {BookingsService} from "./bookings.service";
import {BookingRequest} from "./rest/booking.request";

export {
	BookingsController,
	BookingsService,
	BookingRequest,
};
