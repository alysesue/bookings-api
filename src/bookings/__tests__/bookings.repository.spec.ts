import {BookingsRepository} from "../bookings.repository";
import {DbConnection} from "../../core/db.connection";
import {Booking} from "../../models";
import {Container} from "typescript-ioc";
import {InsertResult, UpdateResult} from "typeorm";

describe("Bookings repository", () => {
	beforeEach(() => {
		jest.resetAllMocks();
	});

	it("should get bookings", async () => {
		Container.bind(DbConnection).to(MockDBConnection);
		MockDBConnection.find.mockImplementation(() => Promise.resolve([]));
		const bookingsRepository = new BookingsRepository();
		const result = await bookingsRepository.getBookings();
		expect(result).toStrictEqual([]);
	});

	it("should save booking", async () => {
		jest.resetAllMocks();
		Container.bind(DbConnection).to(MockDBConnection);
		const insertResult = new InsertResult();
		insertResult.identifiers = [{id: "abc"}];
		MockDBConnection.insert.mockImplementation(() => insertResult);
		const bookingsRepository = new BookingsRepository();
		const booking: Booking = new Booking(new Date(), 60);

		const result = await bookingsRepository.save(booking);
		expect(result.identifiers).toStrictEqual([{id: "abc"}]);
	});

	it('should update booking', async () => {
		jest.resetAllMocks();
		Container.bind(DbConnection).to(MockDBConnection);
		const updateRe = new UpdateResult();

		MockDBConnection.update.mockImplementation(() => updateRe);
		const bookingsRepository = new BookingsRepository();
		const booking: Booking = new Booking(new Date(), 60);

		await bookingsRepository.update(booking);
		expect(MockDBConnection.update.mock.calls[0][1]).toBe(booking);
	});

	it('should get booking', async () => {
		Container.bind(DbConnection).to(MockDBConnection);
		const booking = new Booking(new Date(), 60);
		MockDBConnection.findOne.mockImplementation(() => Promise.resolve(booking));
		const bookingsRepository = Container.get(BookingsRepository);
		const result = await bookingsRepository.getBooking('1');
		expect(result).toStrictEqual(booking);
	});
});

class MockDBConnection extends DbConnection {
	public static insert = jest.fn();
	public static find = jest.fn();
	public static update = jest.fn();
	public static findOne = jest.fn();

	public async getConnection(): Promise<any> {
		const connection = {
			getRepository: () => ({
				find: MockDBConnection.find,
				findOne: MockDBConnection.findOne,
				insert: MockDBConnection.insert,
				update: MockDBConnection.update,
			}),
		};
		return Promise.resolve(connection);
	}
}
