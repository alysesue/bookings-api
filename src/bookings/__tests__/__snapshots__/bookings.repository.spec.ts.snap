// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookings repository should search bookings 1`] = `"{\\"where\\":[{\\"_startDateTime\\":{\\"_type\\":\\"between\\",\\"_value\\":[\\"2020-01-01T14:00:00.000Z\\",\\"2020-01-02T14:00:00.000Z\\"],\\"_useParameter\\":true,\\"_multipleParameters\\":true},\\"_status\\":10}]}"`;

exports[`Bookings repository should search bookings without status 1`] = `"{\\"where\\":[{\\"_startDateTime\\":{\\"_type\\":\\"between\\",\\"_value\\":[\\"2020-01-01T14:00:00.000Z\\",\\"2020-01-02T14:00:00.000Z\\"],\\"_useParameter\\":true,\\"_multipleParameters\\":true}}]}"`;
